str/runners/str_iterative_eh_runner.py

```diff
@@ -49,45 +46,50 @@ def main(
     )
     b = hb.Batch(backend=backend, default_image=os.getenv('DRIVER_IMAGE'))
 
-    external_id_to_cpg_id: dict[str, str] = SampleApi().get_sample_id_map_by_external(
-        dataset, list(external_wgs_ids)
-    )
-    cpg_id_to_external_id = {
-        cpg_id: external_wgs_id
-        for external_wgs_id, cpg_id in external_id_to_cpg_id.items()
-    }
     if dataset in ['tob-wgs', 'hgdp']:
         ref_fasta = (
             'gs://cpg-common-main/references/hg38/v0/Homo_sapiens_assembly38.fasta'
         )
     else:
         ref_fasta = reference_path('broad/ref_fasta')
-    if dataset == 'tob-wgs':
-        analysis_query_model = AnalysisQueryModel(
-            sample_ids=list(external_id_to_cpg_id.values()),
-            projects=[dataset],
-            type=AnalysisType('cram'),
-            status=AnalysisStatus('completed'),
-            meta={'sequencing_type': 'genome', 'source': 'nagim'},
-        )
-    else:
-        analysis_query_model = AnalysisQueryModel(
-            sample_ids=list(external_id_to_cpg_id.values()),
-            projects=[dataset],
-            type=AnalysisType('cram'),
-            status=AnalysisStatus('completed'),
-            meta={'sequencing_type': 'genome'},
-        )
-    crams_path = AnalysisApi().query_analyses(analysis_query_model)
-    cpg_sids_with_crams = set(sid for sids in crams_path for sid in sids['sample_ids'])
-    cpg_sids_without_crams = set(cpg_id_to_external_id.keys()) - cpg_sids_with_crams
-    if cpg_sids_without_crams:
-        external_wgs_sids_without_crams = ', '.join(
-            cpg_id_to_external_id[sid] for sid in cpg_sids_without_crams
-        )
+
+    cram_retrieval_query = gql(
+        """
+        query MyQuery($dataset: String!,$input_cpg_sids: [String!]!) {
+    project(name: $dataset) {
+        sequencingGroups(id: {in_: $input_cpg_sids}) {
+            id
+            sample {
+                externalId
+            }
+            analyses(type: {eq: 'cram'}, active: {eq: true}) {
+                output
+                timestampCompleted
+            }
+        }
+    }
+        """
+    )
+    response = query(
+        cram_retrieval_query,
+        variables={'dataset': dataset, 'input_cpg_sids': input_cpg_sids},
+    )
+    valid_cpg_ids = []
+    crams_path = []
+    for i in response['project']['sequencingGroups']:
+        for cram in i['analyses']:
+            if 'archive' not in cram['output']:  # select the cram that is not archived
+                cram['id'] = i['id']  # ensures cpg id is stored in the dictionary
+                crams_path.append(cram)
+                valid_cpg_ids.append(i['id'])
+                break
+
+    if len(valid_cpg_ids) != len(input_cpg_sids):
+        cpg_sids_without_crams = set(input_cpg_sids) - set(valid_cpg_ids)
```

Comment:
Logically this is fine, you could tidy up the logic a little by using a dictionary instead of maintaining a list of sample IDs and a list of CRAM objects:

```suggestion
    crams_by_id = {}
    for i in response['project']['sequencingGroups']:
        for cram in i['analyses']:
            # ignore archived CRAMs
            if 'archive' in cram['output']:
                continue
            crams_by_id['id'] = cram

    if len(crams_by_id) != len(input_cpg_sids):
        cpg_sids_without_crams = set(input_cpg_sids) - set(crams_by_id.keys())
```


This would then require the change a few lines down:

```
for cpg_sample_id, cram_obj in crams_by_id.items():
    ...
```

You won't gain anything functional, but a dictionary is a decent way to
1. keep the SG ID and corresponding CRAM object linked
2. ensure you only have one CRAM per SG ID without inserting breaks in the logic